packages/dev/core/src/AudioV2/webAudio/subNodes/webAudioBaseSubGraph.ts

```diff
@@ -0,0 +1,75 @@
+import type { Nullable } from "core/types";
+import type { AbstractAudioNode } from "../../abstractAudio/abstractAudioNode";
+import { _AbstractAudioSubGraph } from "../../abstractAudio/subNodes/abstractAudioSubGraph";
+import type { _AbstractAudioSubNode } from "../../abstractAudio/subNodes/abstractAudioSubNode";
+import { _AudioSubNode } from "../../abstractAudio/subNodes/audioSubNode";
+import type { _VolumeAudioSubNode, IVolumeAudioOptions } from "../../abstractAudio/subNodes/volumeAudioSubNode";
+import type { IWebAudioInNode, IWebAudioSubNode, IWebAudioSuperNode } from "../webAudioNode";
+import { _CreateVolumeAudioSubNodeAsync } from "./volumeWebAudioSubNode";
+
+/**
+ * Options for creating a WebAudioBaseSubGraph.
+ */
+export interface IWebAudioBaseSubGraphOptions extends IVolumeAudioOptions {}
+
+/** @internal */
+export abstract class _WebAudioBaseSubGraph extends _AbstractAudioSubGraph {
+    protected _owner: IWebAudioSuperNode;
+    protected _outNode: Nullable<AudioNode> = null;
+
+    /** @internal */
+    public constructor(owner: IWebAudioSuperNode) {
+        super();
+
+        this._owner = owner;
+    }
+
+    /** @internal */
+    public async init(options: Partial<IWebAudioBaseSubGraphOptions>): Promise<void> {
+        this._createAndAddSubNode(_AudioSubNode.VOLUME);
+
+        await this._createSubNodePromisesResolved();
+
+        this.getSubNode<_VolumeAudioSubNode>(_AudioSubNode.VOLUME)?.setOptions(options);
+
+        const volumeNode = this.getSubNode<IWebAudioSubNode>(_AudioSubNode.VOLUME);
+        if (!volumeNode) {
+            return;
+        }
+
+        this._outNode = volumeNode.node;
+
+        // Connect the wrapped downstream WebAudio nodes to the new wrapped WebAudio node.
+        // The wrapper nodes are unaware of this change.
+        if (volumeNode.node && this._downstreamNodes) {
+            const it = this._downstreamNodes.values();
+            for (let next = it.next(); !next.done; next = it.next()) {
+                const inNode = (next.value as IWebAudioInNode).inNode;
+                if (inNode) {
+                    volumeNode.node.connect(inNode);
+                }
+            }
+        }
+    }
+
+    protected abstract readonly _downstreamNodes: Nullable<Set<AbstractAudioNode>>;
+
+    /** @internal */
+    public get inNode(): Nullable<AudioNode> {
+        return this._outNode;
+    }
+
+    /** @internal */
+    public get outNode(): Nullable<AudioNode> {
+        return this._outNode;
+    }
+
+    protected _createSubNode(name: string): Nullable<Promise<_AbstractAudioSubNode>> {
+        switch (name) {
+            case _AudioSubNode.VOLUME:
+                return _CreateVolumeAudioSubNodeAsync(this._owner.engine);
+            default:
+                return null;
+        }
+    }
```

Comment:
This is a weird pattern. Why return a nullable promise? If there is an error, why not throw and use normal promise rejection behavior?

Reply:
It's not an error if this function can't create the specified subnode. The idea is subclasses calling `super._createSubNode` use a `null` return value to know if they should try creating it or not.

See `_WebAudioBusAndSoundSubGraph._createSubNode`:
https://github.com/BabylonJS/Babylon.js/blob/ffbce508d743cad1097c14e00f723fa0b7d596b3/packages/dev/core/src/AudioV2/webAudio/subNodes/webAudioBusAndSoundSubGraph.ts#L58-L73

Reply:
Discussed offline.